File: templates/wasm-node-assemblyscript/examples/math_nodes.ts
/**
 * Math Nodes - Basic arithmetic and mathematical operations
 *
 * This example demonstrates creating nodes for arithmetic operations.
 */

import {
	type ExecutionInput,
	ExecutionResult,
	NodeDefinition,
	PinDefinition,
	parseInput,
	serializeDefinition,
	serializeResult,
} from "../assembly/sdk";

// ============================================================================
// Add Node
// ============================================================================

export function get_add_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "math_add_as";
	def.friendly_name = "Add (AS)";
	def.description = "Adds two numbers together";
	def.category = "Math/Arithmetic";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("a", "A", "First number", "F64").withDefault("0.0"),
	);
	def.addPin(
		PinDefinition.input("b", "B", "Second number", "F64").withDefault("0.0"),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(PinDefinition.output("result", "Result", "Sum of A and B", "F64"));

	return serializeDefinition(def);
}

export function run_add(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const a = getF64Input(input, "a", 0.0);
	const b = getF64Input(input, "b", 0.0);

	const result = ExecutionResult.success();
	result.setOutput("result", (a + b).toString());
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Subtract Node
// ============================================================================

export function get_subtract_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "math_subtract_as";
	def.friendly_name = "Subtract (AS)";
	def.description = "Subtracts B from A";
	def.category = "Math/Arithmetic";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("a", "A", "First number", "F64").withDefault("0.0"),
	);
	def.addPin(
		PinDefinition.input("b", "B", "Second number", "F64").withDefault("0.0"),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(PinDefinition.output("result", "Result", "A minus B", "F64"));

	return serializeDefinition(def);
}

export function run_subtract(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const a = getF64Input(input, "a", 0.0);
	const b = getF64Input(input, "b", 0.0);

	const result = ExecutionResult.success();
	result.setOutput("result", (a - b).toString());
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Multiply Node
// ============================================================================

export function get_multiply_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "math_multiply_as";
	def.friendly_name = "Multiply (AS)";
	def.description = "Multiplies two numbers";
	def.category = "Math/Arithmetic";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("a", "A", "First number", "F64").withDefault("1.0"),
	);
	def.addPin(
		PinDefinition.input("b", "B", "Second number", "F64").withDefault("1.0"),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(PinDefinition.output("result", "Result", "A times B", "F64"));

	return serializeDefinition(def);
}

export function run_multiply(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const a = getF64Input(input, "a", 1.0);
	const b = getF64Input(input, "b", 1.0);

	const result = ExecutionResult.success();
	result.setOutput("result", (a * b).toString());
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Divide Node
// ============================================================================

export function get_divide_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "math_divide_as";
	def.friendly_name = "Divide (AS)";
	def.description = "Divides A by B";
	def.category = "Math/Arithmetic";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("a", "A", "Dividend", "F64").withDefault("0.0"),
	);
	def.addPin(
		PinDefinition.input("b", "B", "Divisor", "F64").withDefault("1.0"),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(PinDefinition.output("result", "Result", "A divided by B", "F64"));
	def.addPin(
		PinDefinition.output(
			"is_valid",
			"Valid",
			"False if division by zero",
			"Bool",
		),
	);

	return serializeDefinition(def);
}

export function run_divide(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const a = getF64Input(input, "a", 0.0);
	const b = getF64Input(input, "b", 1.0);

	const result = ExecutionResult.success();

	if (b == 0.0) {
		result.setOutput("result", "0.0");
		result.setOutput("is_valid", "false");
	} else {
		result.setOutput("result", (a / b).toString());
		result.setOutput("is_valid", "true");
	}

	result.activateExec("exec_out");
	return serializeResult(result);
}

// ============================================================================
// Clamp Node
// ============================================================================

export function get_clamp_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "math_clamp_as";
	def.friendly_name = "Clamp (AS)";
	def.description = "Clamps a value between min and max";
	def.category = "Math/Utility";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("value", "Value", "Value to clamp", "F64").withDefault(
			"0.0",
		),
	);
	def.addPin(
		PinDefinition.input("min", "Min", "Minimum value", "F64").withDefault(
			"0.0",
		),
	);
	def.addPin(
		PinDefinition.input("max", "Max", "Maximum value", "F64").withDefault(
			"1.0",
		),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(PinDefinition.output("result", "Result", "Clamped value", "F64"));

	return serializeDefinition(def);
}

export function run_clamp(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const value = getF64Input(input, "value", 0.0);
	const min = getF64Input(input, "min", 0.0);
	const max = getF64Input(input, "max", 1.0);

	const clamped = Math.max(min, Math.min(max, value));

	const result = ExecutionResult.success();
	result.setOutput("result", clamped.toString());
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Helpers
// ============================================================================

function getF64Input(
	input: ExecutionInput,
	name: string,
	defaultValue: f64,
): f64 {
	if (input.inputs.has(name)) {
		return F64.parseFloat(input.inputs.get(name));
	}
	return defaultValue;
}

File: templates/wasm-node-rust/examples/math_nodes.rs
//! Math Nodes - Basic arithmetic and mathematical operations
//!
//! This example demonstrates creating a multi-node package with
//! arithmetic operations like add, subtract, multiply, divide, and power.

use flow_like_wasm_sdk::*;

// ============================================================================
// Package Definition (Multiple Nodes)
// ============================================================================

package! {
    nodes: [
        {
            name: "math_add",
            friendly_name: "Add",
            description: "Adds two numbers together",
            category: "Math/Arithmetic",
            inputs: {
                exec: Exec,
                a: F64 = 0.0,
                b: F64 = 0.0,
            },
            outputs: {
                exec_out: Exec,
                result: F64,
            },
        },
        {
            name: "math_subtract",
            friendly_name: "Subtract",
            description: "Subtracts B from A",
            category: "Math/Arithmetic",
            inputs: {
                exec: Exec,
                a: F64 = 0.0,
                b: F64 = 0.0,
            },
            outputs: {
                exec_out: Exec,
                result: F64,
            },
        },
        {
            name: "math_multiply",
            friendly_name: "Multiply",
            description: "Multiplies two numbers",
            category: "Math/Arithmetic",
            inputs: {
                exec: Exec,
                a: F64 = 0.0,
                b: F64 = 0.0,
            },
            outputs: {
                exec_out: Exec,
                result: F64,
            },
        },
        {
            name: "math_divide",
            friendly_name: "Divide",
            description: "Divides A by B",
            category: "Math/Arithmetic",
            inputs: {
                exec: Exec,
                a: F64 = 0.0,
                b: F64 = 1.0,
            },
            outputs: {
                exec_out: Exec,
                result: F64,
                is_valid: Bool,
            },
        },
        {
            name: "math_power",
            friendly_name: "Power",
            description: "Raises A to the power of B",
            category: "Math/Arithmetic",
            inputs: {
                exec: Exec,
                base: F64 = 0.0,
                exponent: F64 = 1.0,
            },
            outputs: {
                exec_out: Exec,
                result: F64,
            },
        },
        {
            name: "math_clamp",
            friendly_name: "Clamp",
            description: "Clamps a value between min and max",
            category: "Math/Utility",
            inputs: {
                exec: Exec,
                value: F64 = 0.0,
                min: F64 = 0.0,
                max: F64 = 1.0,
            },
            outputs: {
                exec_out: Exec,
                result: F64,
            },
        }
    ]
}

// ============================================================================
// Node Implementations
// ============================================================================

run_package!(run_node);

fn run_node(node_name: &str, mut ctx: Context) -> ExecutionResult {
    match node_name {
        "math_add" => {
            let a = ctx.get_f64("a").unwrap_or(0.0);
            let b = ctx.get_f64("b").unwrap_or(0.0);
            ctx.set_output("result", a + b);
            ctx.success()
        }
        "math_subtract" => {
            let a = ctx.get_f64("a").unwrap_or(0.0);
            let b = ctx.get_f64("b").unwrap_or(0.0);
            ctx.set_output("result", a - b);
            ctx.success()
        }
        "math_multiply" => {
            let a = ctx.get_f64("a").unwrap_or(0.0);
            let b = ctx.get_f64("b").unwrap_or(0.0);
            ctx.set_output("result", a * b);
            ctx.success()
        }
        "math_divide" => {
            let a = ctx.get_f64("a").unwrap_or(0.0);
            let b = ctx.get_f64("b").unwrap_or(1.0);

            if b == 0.0 {
                ctx.set_output("result", 0.0);
                ctx.set_output("is_valid", false);
                ctx.warn("Division by zero");
            } else {
                ctx.set_output("result", a / b);
                ctx.set_output("is_valid", true);
            }
            ctx.success()
        }
        "math_power" => {
            let base = ctx.get_f64("base").unwrap_or(0.0);
            let exponent = ctx.get_f64("exponent").unwrap_or(1.0);
            ctx.set_output("result", base.powf(exponent));
            ctx.success()
        }
        "math_clamp" => {
            let value = ctx.get_f64("value").unwrap_or(0.0);
            let min = ctx.get_f64("min").unwrap_or(0.0);
            let max = ctx.get_f64("max").unwrap_or(1.0);
            ctx.set_output("result", value.max(min).min(max));
            ctx.success()
        }
        _ => ctx.fail(format!("Unknown node: {}", node_name)),
    }
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    #[test]
    fn test_add() {
        assert_eq!(5.0 + 3.0, 8.0);
    }

    #[test]
    fn test_divide_by_zero() {
        let b = 0.0f64;
        assert!(b == 0.0);
    }

    #[test]
    fn test_power() {
        assert_eq!(2.0f64.powf(3.0), 8.0);
    }

    #[test]
    fn test_clamp() {
        let value = 15.0f64;
        let result = value.max(0.0).min(10.0);
        assert_eq!(result, 10.0);
    }
}
